File: D. Stack/stack.cpp
#include <iostream>
#include <new>
using namespace std;

// Stack
template <typename T>
class Stack {
  public:
    class Node;
  private:
    int size = 0;
    Node *head = NULL;
    Node *tail = NULL;
  public:
    // Node class
    class Node {
      public:
        T data;
        Node *next;
        // Constructor
        Node(T data = NULL, Node *next = NULL) {
          this->data = data;
          this->next = next;
        }
    };

    // Push an element to this Stack, O(1)
    void push(T elem) {
      if (size == 0) {
        head = tail = new Node(elem);
      } else {
        Node *temp = new Node(elem, head);
        head = temp;
      }
      size++;
    }

    // Pop an element from this Stack. O(1)
    void pop() {
      Node *temp = head;
      head = head->next;
      delete temp;
      size--;
    }

    // Check the value of the first node if it exists, O(1)
    T peekFirst() {
      try {if (size == 0) throw 101;} catch(...) {cout<<"Empty list";}
      return head.data;
    }

    // Find the index of a particular value in the linked list, O(n)
    int indexOf(T value) {
      int index = 0;
      Node *trav = head;
      for (; trav != NULL; trav = trav->next, index++) {
        if (value == trav->data) {
          return index;
        }
      }
      return -1;
    }

    // Check is a value is contained within the linked list
    bool contains(T value) {
      return indexOf(value) != -1;
    }

    // Print values in linked list
    void print() {
      Node *trav = head;
      while (trav != NULL) {
        cout<<trav->data<<" ";
        trav = trav->next;
      }
    }
};

int main()
{
  Stack<int> A;
  A.push(4);
  A.push(3);
  A.push(2);
  A.push(1);
  A.push(0);
  A.push(-1);
  A.pop();
  A.pop();
  A.print();
  return 0;
}

File: Stack/stack.py
class Stack:
    # Node class inside stack
    class Node:
        # Node constructor
        def __init__(self, data = None, next = None):
            self.data = data
            self.next = next

    # Stack constructor
    def __init__(self):
        self.head = None
        self.size = 0

    # Push element at head in stack, O(1)
    def push(self, data):
        if self.head == None:
            self.head = self.Node(data)
        else:
            temp = self.Node(data, self.head)
            self.head = temp
        self.size += 1

    # Pop element from stack at top, O(1)
    def pop(self):
        temp = self.head
        self.head = self.head.next
        del temp
        self.size -= 1

    # Returns size of linked list, O(1)
    def getSize(self):
        return self.size

    # Find the index of a particular value in stack, O(n)
    def indexOf(self, data):
        index = 0
        temp = self.head
        while index < self.size:
            if temp.data == data:
                return index
            temp = temp.next
            index += 1
        else:
            return -1

    # Check is a value is contained within the stack
    def contains(self, data):
        return self.indexOf(data) != -1

    # Print stack
    def show(self):
        temp = self.head
        print('Stack:', end = ' ')
        while temp != None:
            print(temp.data, end = ' ')
            temp = temp.next

x = Stack()
x.push(2)
x.push(3)
x.push(4)
x.push(6)
x.push(7)
x.pop()
x.pop()
x.show()
